sound/soc/intel/boards/sdw_rt711_rt1308_rt715.c

```diff
@@ -364,83 +366,263 @@ struct codec_dai dai_lists[] = {
 	},
 };
 
-struct snd_soc_dai_link dailink[] = {
-	{
-		.name = "SDW0-Playback",
-		.id = 0,
-		.init = headset_init,
-		.no_pcm = 1,
-		.dpcm_playback = 1,
-		.nonatomic = true,
-		SND_SOC_DAILINK_REG(sdw0_pin2, sdw0_codec, platform),
-	},
-	{
-		.name = "SDW0-Capture",
-		.id = 1,
-		.no_pcm = 1,
-		.dpcm_capture = 1,
-		.nonatomic = true,
-		SND_SOC_DAILINK_REG(sdw0_pin3, sdw0_codec, platform),
-	},
-	{
-		.name = "SDW1-Playback",
-		.id = 2,
-		.no_pcm = 1,
-		.dpcm_playback = 1,
-		.nonatomic = true,
-		SND_SOC_DAILINK_REG(sdw1_pin2, sdw1_codec, platform),
-	},
-	{
-		.name = "SDW3-Capture",
-		.id = 4,
-		.no_pcm = 1,
-		.dpcm_capture = 1,
-		.nonatomic = true,
-		SND_SOC_DAILINK_REG(sdw3_pin2, sdw3_codec, platform),
-	},
+static int get_sdw_dai_link_num(struct snd_soc_acpi_mach_params *mach_params)
+{
+	const struct snd_soc_acpi_link_adr *link;
+	int num = 0;
+
+	link = mach_params->links;
```

Comment:
didn't we agree that the links are optional if you can match with just the link_mask, so you would need to test if link != NULL before de-referencing?

Reply:
yes, I got it.

Reply:
> didn't we agree that the links are optional if you can match with just the link_mask, so you would need to test if link != NULL before de-referencing?

The dai links seems to be created dynamically, we will need a predefined codecs which will let the machine driver look messy. So, I would suggest that we can assume that we will use another machine driver if links is NULL. So we will return error and show an error message if it is NULL in the machine driver.

Reply:
@plbossart if links == NULL, do we need to do it as original machine driver ?  So I should not delete the predefined codecs lists ? Thanks!

Reply:
Sorry, I don't understand the question.

Reply:
> didn't we agree that the links are optional if you can match with just the link_mask, so you would need to test if link != NULL before de-referencing?

@plbossart     if (link == NULL), just return error ?  it is impossible to generate dai dynamically.
Or go back with predefine dai link way ?